### PR TITLE
Add README.md for JSInterop

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Part of https://github.com/dotnet/aspnetcore/issues/27561

### CODE REVIEW

File: src/JSInterop/README.md [hunk 1]
```diff
@@ -1,8 +1,47 @@
-# jsinterop
+# JSInterop
 
-This repo is for `Microsoft.JSInterop`, a package that provides abstractions and features for interop between .NET and JavaScript code.
+This directory contains sources for `Microsoft.JSInterop`, a package that provides abstractions and features for interop between .NET and JavaScript code.
```
Comment (guardrex) [inline src/JSInterop/README.md:3]:
```suggestion
This directory contains sources for [`Microsoft.JSInterop`](https://www.nuget.org/packages/Microsoft.JSInterop), a package that provides abstractions and features for interop between .NET and JavaScript code.
```

File: src/JSInterop/README.md [hunk 2]
```diff
@@ -1,8 +1,47 @@
...
+
+This section provides a brief overview of the architecture.
+
+`Microsoft.JSInterop` is a .NET package with the following roles:
+
+ * Defining abstractions to describe how .NET code can invoke JavaScript code and pass parameters. These abstractions include `IJSRuntime`, `IJSInProcessRuntime`, `DotNetObjectReference`, `IJSObjectReference`, and others.
+ * Providing platform-independent abstract base class implementations of those abstractions, such as `JSRuntime` and `JSObjectReference`. These implement common logic around handling errors and asynchrony, even though they are independent of any particular runtime environment.
+ * Providing extension methods on `IJSRuntime` that simplify making calls with differing numbers of parameters, cancellation tokens, etc.
+
+For these types to become usable in a particular runtime environment such as Blazor Server or Blazor WebAssembly, the runtime environment implements its own concrete subclasses that know how to dispatch calls to the actual JavaScript runtime that is available in that environment. For example, Blazor Server uses the SignalR-based circuit to send invocations to the end user's browser.
```
Comment (guardrex) [inline src/JSInterop/README.md:20]:
```suggestion
For these types to become usable in a particular runtime environment, such as Blazor Server or Blazor WebAssembly, the runtime environment implements its own concrete subclasses that know how to dispatch calls to the actual JavaScript runtime that is available in that environment. For example, Blazor Server uses the SignalR-based circuit to send invocations to the end user's browser.
```

File: src/JSInterop/README.md [hunk 3]
```diff
@@ -1,8 +1,47 @@
...
+
+`Microsoft.JSInterop.JS` is the JavaScript-side counterpart to the above. It runs within a standard web browser environment, receives the invocations from .NET code, executes them, and sends back results in the format understood by the `JSRuntime` base class. This includes special handling for certain parameter types such as `ElementReference` and `DotNetObjectReference`. `Microsoft.JSInterop.JS` also exposes JavaScript functions that can be used to issue calls from JavaScript to .NET.
```
Comment (guardrex) [inline src/JSInterop/README.md:22]:
```suggestion
`Microsoft.JSInterop.JS` is the JavaScript-side counterpart to the preceding. It runs within a standard web browser environment, receives the invocations from .NET code, executes them, and sends back results in the format understood by the `JSRuntime` base class. This includes special handling for certain parameter types such as `ElementReference` and `DotNetObjectReference`. `Microsoft.JSInterop.JS` also exposes JavaScript functions that can be used to issue calls from JavaScript to .NET.
```

File: src/JSInterop/README.md [hunk 4]
```diff
@@ -1,8 +1,47 @@
...
 
+## Description
+
+This section provides a brief overview of the architecture.
+
+`Microsoft.JSInterop` is a .NET package with the following roles:
+
+ * Defining abstractions to describe how .NET code can invoke JavaScript code and pass parameters. These abstractions include `IJSRuntime`, `IJSInProcessRuntime`, `DotNetObjectReference`, `IJSObjectReference`, and others.
+ * Providing platform-independent abstract base class implementations of those abstractions, such as `JSRuntime` and `JSObjectReference`. These implement common logic around handling errors and asynchrony, even though they are independent of any particular runtime environment.
+ * Providing extension methods on `IJSRuntime` that simplify making calls with differing numbers of parameters, cancellation tokens, etc.
```
Comment (guardrex) [inline src/JSInterop/README.md:18]:
```suggestion
 * Providing extension methods on `IJSRuntime` that simplify making calls with differing numbers of parameters, cancellation tokens, and other characteristics.
```

File: src/JSInterop/README.md [hunk 5]
```diff
@@ -1,8 +1,47 @@
...
+
+Since `Microsoft.JSInterop.JS` is platform-independent, runtime environments such as Blazor Server or Blazor WebAssembly must initialize it by registering environment-specific callbacks that know how to dispatch invocations across their own communication channels.
+
+## Development Setup
+
+### Build and test
+
+To build the .NET code, you can:
+
+ * Run `dotnet build` in the `Microsoft.JSInterop/src` directory
```
Comment (guardrex) [inline src/JSInterop/README.md:32]:
```suggestion
 * Run `dotnet build` in the `Microsoft.JSInterop/src` directory.
```

File: src/JSInterop/README.md [hunk 6]
```diff
@@ -1,8 +1,47 @@
...
+ * Run `dotnet build` or `dotnet test` in the `Microsoft.JSInterop/test` directory. You can also read more about how to [run the tests on the command line](../../docs/BuildFromSource.md#running-tests-on-command-line).
+
+Alternatively, open `JSInterop.slnf` in Visual Studio.
+
+To build the JavaScript code,
```
Comment (guardrex) [inline src/JSInterop/README.md:37]:
```suggestion
To build the JavaScript code, execute the following commands in a command shell:
```

**Review by guardrex — APPROVED** [review]
Very minor nits.

I'm not applying all of the tiny little doc concepts (e.g., 2nd person and future tense, pronouns, API doc cross-links, unpaired heading, etc.). These suggestions are just the *better nits* for a README. :smile:

**Comment by SteveSandersonMS** [comment]
Thanks @guardrex!

File: src/JSInterop/README.md [hunk 7]
```diff
@@ -1,8 +1,47 @@
...
+
+Alternatively, open `JSInterop.slnf` in Visual Studio.
+
+To build the JavaScript code, execute the following commands in a command shell:
+
+ * `cd Microsoft.JSInterop.JS/src`
+ * `npm run preclean`
+ * `npm run build`
+
+**Warning:** Due to special requirements related to ASP.NET Core's CI process for Linux distributions, we store the compiled JavaScript artifacts for `Microsoft.JSInterop.JS` in source control in the `Microsoft.JSInterop.JS\src\dist` directory. If you edit and build JavaScript sources, your Git client should indicate that those outputs have changed. You will need to include changes to those `dist` files in any PRs that you submit. When this leads to merge conflicts, we have to resolve them manually by rebasing and rebuilding.
```
Comment (SteveSandersonMS) [inline src/JSInterop/README.md:43]:
```suggestion
**Warning:** Due to special requirements related to ASP.NET Core's CI process for Linux distributions, we store the compiled JavaScript artifacts for `Microsoft.JSInterop.JS` in source control in the `Microsoft.JSInterop.JS/src/dist` directory. If you edit and build JavaScript sources, your Git client should indicate that those outputs have changed. You will need to include changes to those `dist` files in any PRs that you submit. When this leads to merge conflicts, we have to resolve them manually by rebasing and rebuilding.
```

File: src/JSInterop/README.md [hunk 8]
```diff
@@ -1,8 +1,47 @@
...
+
+Since `Microsoft.JSInterop.JS` is platform-independent, runtime environments such as Blazor Server or Blazor WebAssembly must initialize it by registering environment-specific callbacks that know how to dispatch invocations across their own communication channels.
+
+## Development Setup
+
+### Build and test
+
+To build the .NET code, you can:
+
+ * Run `dotnet build` in the `Microsoft.JSInterop/src` directory.
```
Comment (captainsafia) [inline src/JSInterop/README.md:32]:
```suggestion
 * Run `dotnet build` in the `Microsoft.JSInterop/src` directory. You can also read more [on building a subset of the code](../../docs/BuildFromSource.md#building-a-subset-of-the-code).
```

**Review by captainsafia — APPROVED** [review]

**Pull request merged**